bpf/tests/eni_nlb_symetric_routing_host.c

```diff
@@ -0,0 +1,203 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/* Copyright Authors of Cilium */
+
+#include "common.h"
+
+/* Enable CT debug output */
+#undef QUIET_CT
+
+#define DEBUG
+
+#include <bpf/ctx/skb.h>
+#include "pktgen.h"
+
+/* Enable code paths under test*/
+#define ENABLE_IPV4
+#define ENABLE_NODEPORT
+#define ENABLE_ROUTING
+
+#define PRIMARY_IFACE 1
+#define SECONDARY_IFACE 2
+#define BACKEND_IFACE 123
+
+#define NATIVE_DEV_IFINDEX PRIMARY_IFACE
+
+#define BACKEND_EP_ID 127
+
+#define DEFAULT_ROUTE_MAC mac_one
+#define NODE_MAC mac_two
+#define LOCAL_BACKEND_MAC mac_three
+
+#define ctx_redirect mock_ctx_redirect
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__type(key, __u32);
+	__type(value, __u32);
+	__uint(max_entries, 1);
+} redirect_ifindex_map __section_maps_btf;
+
+static __always_inline __maybe_unused int
+mock_ctx_redirect(const struct __sk_buff *ctx __maybe_unused,
+		  int ifindex __maybe_unused, __u32 flags __maybe_unused)
+{
+	__u32 key = 0;
+	__u32 *value = map_lookup_elem(&redirect_ifindex_map, &key);
+
+	if (value)
+		*value = ifindex;
+
+	return CTX_ACT_REDIRECT;
+}
+
+#include <bpf_host.c>
+
+#include "lib/endpoint.h"
+#include "lib/ipcache.h"
+
+#define TO_NETDEV	0
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PROG_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(max_entries, 1);
+	__array(values, int());
+} entry_call_map __section(".maps") = {
+	.values = {
+		[TO_NETDEV] = &cil_to_netdev,
+	},
+};
+
+/* Setup for this test:
+ *
+ * +-------World IP---------+    +----------Pod IP---------+
+ * | v4_ext_one:tcp_src_one | -> | v4_pod_one:tcp_dst_one |
+ * +------------------------+    +------------------------+
+ *          ^                            |
+ *          \---------------------------/
+ *
+ * This happens when in ENI mode and we get traffic via a NLB in ip-mode + preserve-client-ip.
+ * NLB NAT'ing is invisible to us, so we effectively see world to pod IP traffic.
+ * We need to make sure that in KPR mode, we do not masquerade the pod IP for the return traffic
+ * and route it out of the interface associated with the pod IP, even though egress traffic to
+ * the world is normally routed via the primary interface.
```

Comment:
I've seen "KPR mode" referenced a few times when this issue was discussed. While my understanding is that it's actually about the masquerading mode - iptables vs BPF?

Reply:
I guess what I'm saying is that if this is purely about the masquearding mode, then maybe this code section should even be guarded by `ENABLE_MASQUERADE_IPV4`.

Reply:
We can. As extra context, we have the same asymmetric routing issue in Cilium iptables mode as well, but we decided to only fix it for "KPR". The other way to go about this would require iptables + conntrack + skb marks + routes, which is just a pain in the neck.

So, essentially, if we have BPF installed on the to-netdev devices we want to do this. I would have to lookup under which flags this happens.

Technically, this is only the routing part, the "no not masquerade reply traffic from a pod" already works, at least in BPF mode. If we somehow can have situations where we do have to-netdev programs installed by still use iptables for masquerading then things end up half broken. At least for the situation I am trying to fix

Reply:
>If we somehow can have situations where we do have to-netdev programs installed by still use iptables for masquerading then things end up half broken. At least for the situation I am trying to fix

We definitely have such situations :). As examples - `to-netdev` is used by Wireguard and HostFW. Neither require BPF-masq.

(I'd want to emphasize the difference between "KPR" and "BPF-masq". Right now they're a bit intermixed because BPF-masq depends on `ENABLE_NODEPORT`. But I'd really want to change that, and we shouldn't assume that KPR and BPF-masq always go together.)